<!DOCTYPE html>
<html>
<head>
	<meta http-equiv="Cache-control" content="no-cache">
	<meta http-equiv="Expires" content="-1">
	<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
	<meta name="pdfkit-page-size" content="Legal"/>
    <meta name="pdfkit-orientation" content="Landscape"/>
	<link rel="stylesheet" href="/static/css/bootstrap.min.css" >
	<style>
		*{
			font-family: 'Roboto', sans-serif;
			font-size: 12px;
			margin: 0;
			padding: 0;
		}

		section {
			padding: 10px 0px 0px 10px;
		}

		table { page-break-after:auto }
		tr    { page-break-inside:avoid; page-break-after:auto }
		td    { page-break-inside:avoid; page-break-after:auto }

		.table-bordered tr th, .table-bordered tr td {
			border: 1px solid #000;
		}

		/* .hz-line {
			height:1px;
			border-width:0;
			color:#868B8E;
			background-color:#868B8E;
			width: 50%;
		} */

		hr {
			page-break-inside:avoid; page-break-after:auto 
		}

		h5 {
			font-weight: bold;
			padding-bottom: 5px;
			color: #B43218;
			font-size: 14px;
		}

		.technical-validation{
			width: 100%;
		}

		.technical-validation table tr th{
			font-size: 12px;
		}

		.technical-validation thead tr th{
			background-color: #F2F1F0;
			text-align: center;
			color: #000000;
		}


		.technical-validation table tr td{
			font-size: 10px;
            color:#6b6b6b;
            font-weight:800;
		}

		.genomic-alteration {
			width: 100%;
		}

		.genomic-alteration h5, .gene-panel h5 {
			padding-bottom: 10px;
			text-transform: uppercase;
		}

		.genomic-alteration table tr th {
			background-color: #F2F1F0;
			color: #000000;
			font-size: 12px;
			text-transform: uppercase;
		}

		.genomic-alteration table tr td {
			color : #6b6b6b;
            font-weight:800;
            font-size: 10px;
			text-transform:capitalize;
		}

		.genomic-alteration table tr td.tot-title {
			color: #000000;
            font-weight: 800;
			text-align: center;
			text-transform: uppercase;
		}

		.gene-panel {
			font-size: 12px;
		}

		.gene-panel ul {
			list-style: none;
			display: block;
		}
		
		.gene-panel ul li {
			display: inline-block;
			height: 28px;
			width: 95px;
			line-height: 25px;
			margin: 2px;
			/*background: #657786;
			color: #ffffff; */
			border: 1px solid #B9B7BD;
			color: #000000;
			font-size: 10px;
			text-align: center;
			font-weight: 500;
		}

	</style>
</head>
<body>
	<section class="technical-validation">
		<h5 class="font-weight-bold">TECHNICAL VALIDATION</h5>
		<table class="table table-bordered">
			<thead id="tech_val_table_header">

			</thead>
			<tbody id="tech_val_table_data">
				
			</tbody>
		</table>
	</section>
	<section class="genomic-alteration">
		<h5 class="font-weight-bold">Categorization of analyzed genomic alterations</h5>
		<table class="table table-bordered table-hover">
			<tr>
				<th colspan="2">Small Somatic and Germline Alterations</th>
			</tr>
			<tr>
				<td>ALL CODING EXONS</td>
				<td>52 genes</td>
			</tr>
			<tr>
				<td>HOTSPOTS</td>
				<td>27 genes</td>
			</tr>
			<tr>
				<th colspan="2">PHARMACOGENETIC VARIANTS</th>
			</tr>
			<tr>
				<td>SNPs</td>
				<td>4 genes</td>
			</tr>
			<tr>
				<th colspan="2">COPY-NUMBER ALTERATIONS</th>
			</tr>
			<tr>
				<td>TILED SNP FOR GENOME-WIDE CNV</td>
				<td>3128 SNPs</td>
			</tr>
			<tr>
				<td>DIRECTED ANALYSIS TO INCREASE SENSITIVTY</td>
				<td>23 genes</td>
			</tr>
			<tr>
				<th colspan="2">STRUCTURAL VARIATION</th>
			</tr>
			<tr>
				<td>GENE FUSION BY INTRONIC SEQUENCING</td>
				<td>3 genes</td>
			</tr>
			<tr>
				<td>GENE-BODY SEQUENCING (E.G. BRCA1/2)</td>
				<td>4 genes</td>
			</tr>
			<tr>
				<th colspan="2">MICROSATELLITE INSTABILITY & TMB</th>
			</tr>
			<tr>
				<td>MICROSATELLITES</td>
				<td>63</td>
			</tr>
			<tr>
				<td>HYPERMUTATION, ENTIRE DESIGN FOOTPRINT</td>
				<td>YES</td>
			</tr>
			<tr>
				<td>ASSOCIATED GENES</td>
				<td>6</td>
			</tr>
			<tr>
				<th colspan="2">DNA REPAIR DEFICIENCY</th>
			</tr>
			<tr>
				<td>ASSOCIATED GENES</td>
				<td>18</td>
			</tr>
			<tr>
				<td class="tot-title">Total Size</td>
				<td class="font-weight-bold">1.34 Mb</td>
			</tr>

		</table>
	</section>
	<section class="gene-panel">
		<h5 class="font-weight-bold">Exonic regions sequenced to detect small somatic and germline alterations</h5>
		<ul>
			<li>AKT1</li>
			<li>BRAF</li>
			<li>CCND1</li>
			<li>CDK4</li>
			<li>CDK6</li>
			<li>CDKN1A</li>
			<li>CDKN2A</li>
			<li>CTNNB1</li>
			<li>CUL3</li>
			<li>DICER1</li>
			<li>FOXO1</li>
			<li>HRAS</li>
			<li>IDH1</li>
			<li>KRAS</li>
			<li>MED12</li>
			<li>NRAS</li>
			<li>PIK3CA</li>
			<li>PIK3CB</li>
			<li>PIK3CD</li>
			<li>PIK3R1</li>
			<li>PIK3R2</li>
			<li>POLD1</li>
			<li>POLE</li>
			<li>SF3B1</li>
			<li>SPOP</li>
			<li>U2AF1</li>
			<li>XPO1</li>
			<li>APC</li>
			<li>AR</li>
			<li>ATM</li>
			<li>BRCA1</li>
			<li>BRCA2</li>
			<li>CDK12</li>
			<li>FOXA1</li>
			<li>KDM6A</li>
			<li>KMT2A</li>
			<li>KMT2C</li>
			<li>KMT2D</li>
			<li>PALB2</li>
			<li>PTEN</li>
			<li>RB1</li>
			<li>TP53</li>
			<li>ARID1A</li>
			<li>ARID2</li>
			<li>ATR</li>
			<li>BARD1</li>
			<li>BRIP1</li>
			<li>CDH1</li>
			<li>CDKN1B</li>
			<li>CHD1</li>
			<li>CHEK2</li>
			<li>FANCA</li>
			<li>JAK1</li>
			<li>KEAP1</li>
			<li>MET</li>
			<li>MGA</li>
			<li>MLH1</li>
			<li>MLH3</li>
			<li>MRE11A</li>
			<li>MSH2</li>
			<li>MSH3</li>
			<li>MSH6</li>
			<li>NBN</li>
			<li>NCOR1</li>
			<li>NKX3-1</li>
			<li>PMS1</li>
			<li>PMS2</li>
			<li>RAD50</li>
			<li>RAD51</li>
			<li>RAD51B</li>
			<li>RAD51C</li>
			<li>RAD51D</li>
			<li>RNF43</li>
			<li>SETD2</li>
			<li>SPEN</li>
			<li>ZFHX3</li>
			<li>ZMYM3</li>
			<li>RAD54L</li>
			<li>PIK3R1</li>
		</ul>
	</section>
	<hr class="hz-line">
	<section class="gene-panel">
		<h5 class="font-weight-bold">Intronic and intergenic regions sequenced to increase sensitivity to detect copy-number alterations</h5>
		<ul>
			<li>AR</li>
			<li>AKT1</li>
			<li>AR_ENHANCER</li>
			<li>ATM</li>
			<li>BRCA2</li>
			<li>CCND1</li>
			<li>CDK12</li>
			<li>CDKN2A</li>
			<li>CDKN2B</li>
			<li>CHD1</li>
			<li>CHEK2</li>
			<li>ERG</li>
			<li>FANCA</li>
			<li>MYC</li>
		    <li>NKX3-1</li>
			<li>PIK3CA</li>
			<li>PIK3CB</li>
			<li>PIK3R1</li>
			<li>PTEN</li>
			<li>RB1</li>
			<li>TMPRSS2</li>
			<li>TP53</li>
			<li>ZBTB16</li>
		</ul>
	</section>
	<hr class="hz-line">
	<section class="gene-panel">
		<h5 class="font-weight-bold">Intronic regions sequenced to detect fusions and structural rearrangements</h5>
		<ul>
			<li>AR</li>
			<li>BRCA2</li>
			<li>ERG</li>
			<li>ETV1</li>
			<li>PTEN</li>
			<li>RB1</li>
			<li>TMPRSS2</li>
		</ul>
	</section>
	<hr class="hz-line">
	<section class="gene-panel">
		<h5 class="font-weight-bold">Pharmacogenomic SNPs analysed</h5>
		<ul>
			<li>COMT</li>
			<li>DPYD</li>
			<li>HSD3B1</li>
			<li>HOXB13</li>
		</ul>
	</section>
</body>
</html>
