<!DOCTYPE html>
<html>
<head>
	<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
	<meta name="pdfkit-page-size" content="Legal"/>
    <meta name="pdfkit-orientation" content="Landscape"/>
	<link rel="stylesheet" href="/nfs/IPCM/script/pdf/static/css/bootstrap.min.css" >
	<style>
		*{
			font-family: 'Roboto', serif;
			font-size: 14px;
		}

		section {
			padding: 10px 0px 10px 10px;
		}

		table { page-break-after:auto }
		tr    { page-break-inside:avoid; page-break-after:auto }
		td    { page-break-inside:avoid; page-break-after:auto }

		.table-bordered tr th, .table-bordered tr td {
			border: 1px solid #000;
		}

		/* .hz-line {
			height:1px;
			border-width:0;
			color:#868B8E;
			background-color:#868B8E;
			width: 50%;
		} */

		h5 {
			font-weight: bold;
			padding-bottom: 15px;
			color: #B43218;
			font-size: 16px;
		}

		.technical-validation{
			width: 100%;
			padding: 25px 0px;
		}

		.technical-validation table tr th{
			font-size: 14px;
		}

		.technical-validation thead tr th{
			background-color: #d3d3d3;
			text-align: center;
			color: #000000;
		}

		.technical-validation table tr td{
			font-size: 13px;
            /* color :#7E7C73; */
            color:#6b6b6b;
            font-weight:800;
		}
		
		.genomic-alteration {
			width: 100%;
		}
		.genomic-alteration h5, .gene-panel h5 {
			padding-bottom: 10px;
			text-transform: uppercase;
		}

		.genomic-alteration table tr th {
			background-color: #d3d3d3;
			color: #000000;
			font-size: 14px;
			text-transform: uppercase;
		}

		.genomic-alteration table tr td {
			/* color :#7E7C73; */
            color : #6b6b6b;
            font-weight:800;
			font-size: 13px;
			text-transform:capitalize;
		}

		.genomic-alteration table tr td.tot-title {
			color: #000000;
            font-weight: 800;
            text-align:center;
			text-transform: uppercase;
		}

		.gene-panel {
			font-size: 14px;
		}

		.gene-panel ul {
			list-style: none;
			display: block;
		}
		
		.gene-panel ul li {
			display: inline-block;
			height: 25px;
			width: 100px;
			line-height: 30px;
			margin: 5px;
			background: #657786;
			color: #ffffff;
			font-size: 12px;
			text-align: center;
			font-weight: 600;
		}

	</style>
</head>
<body>
	<section class="technical-validation">
		<h5 class="font-weight-bold">TECHNICAL VALIDATION</h5>
		<table class="table table-bordered">
			<thead id="tech_val_table_header">

			</thead>
			<tbody id="tech_val_table_data">
				
			</tbody>
		</table>
	</section>
	<section class="genomic-alteration">
		<h5 class="font-weight-bold">Categorization of analyzed genomic alterations</h5>
		<table class="table table-bordered table-hover">
			<tr>
				<th colspan="2">Small Somatic and Germline Alterations</th>
			</tr>
			<tr>
				<td>ALL CODING EXONS</td>
				<td>197 genes</td>
			</tr>
			<tr>
				<td>HOTSPOTS</td>
				<td>132 genes</td>
			</tr>
			<tr>
				<th colspan="2">PHARMACOGENETIC VARIANTS</th>
			</tr>
			<tr>
				<td>SNPs</td>
				<td>9 genes</td>
			</tr>
			<tr>
				<th colspan="2">COPY-NUMBER ALTERATIONS</th>
			</tr>
			<tr>
				<td>TILED SNP FOR GENOME-WIDE CNV</td>
				<td>2814 SNPs</td>
			</tr>
			<tr>
				<td>INCREASED TILING FOR CNV RESOLUTION</td>
				<td>86 genes</td>
			</tr>
			<tr>
				<th colspan="2">STRUCTURAL VARIATION</th>
			</tr>
			<tr>
				<td>GENE FUSION BY INTRONIC SEQUENCING</td>
				<td>19 genes</td>
			</tr>
			<tr>
				<td>GENE-BODY SEQUENCING</td>
				<td>9 genes</td>
			</tr>
			<tr>
				<th colspan="2">MICROSATELLITE INSTABILITY & TMB</th>
			</tr>
			<tr>
				<td>MICROSATELLITES</td>
				<td>63</td>
			</tr>
			<tr>
				<td>HYPERMUTATION, ENTIRE DESIGN FOOTPRINT</td>
				<td>YES</td>
			</tr>
			<tr>
				<td>ASSOCIATED GENES</td>
				<td>7</td>
			</tr>
			<tr>
				<td class="tot-title">Total Size</td>
				<td class="font-weight-bold">2.4 Mb</td>
			</tr>

		</table>
	</section>
	<section class="gene-panel">
		<h5 class="font-weight-bold">Exonic regions sequenced to detect clonal hematopoiesis</h5>
		<ul>
			<li>BCL2</li>
			<li>CARD11</li>
			<li>CBFB</li>
			<li>CD79B</li>
			<li>CEBPA</li>
			<li>DNMT3A</li>
			<li>FLT3</li>
			<li>IKZF1</li>
			<li>MYD88</li>
			<li>NSD1</li>
			<li>PTPN11</li>
			<li>RARA</li>
			<li>RUNX1</li>
			<li>SF3B1</li>
			<li>SRSF2</li>
			<li>TET2</li>
			<li>WHSC1/NSD2</li>
			<li>XPO1</li>
		</ul>
	</section>
	<hr class="hz-line">
	<section class="gene-panel">
		<h5 class="font-weight-bold">Exonic regions sequenced to detect small somatic and germline alterations</h5>
		<ul>
			<li>ACVR1</li>
			<li>ACVR1B</li>
			<li>ACVR2A</li>
			<li>AKT1</li>
			<li>AKT2</li>
			<li>AKT3</li>
			<li>ALB</li>
			<li>ALK</li>
			<li>AMER1</li>
			<li>ANKRD11</li>
			<li>APC</li>
			<li>AR</li>
			<li>ARAF</li>
			<li>ARHGAP35</li>
			<li>ARID1A</li>
			<li>ARID1B</li>
			<li>ARID2</li>
			<li>ARID5B</li>
			<li>ASXL1</li>
			<li>ASXL2</li>
			<li>ATM</li>
			<li>ATR</li>
			<li>ATRX</li>
			<li>AXIN1</li>
			<li>AXIN2</li>
			<li>AXL</li>
			<li>B2M</li>
			<li>BAP1</li>
			<li>BARD1</li>
			<li>BCL10</li>
			<li>BCL2L11</li>
			<li>BCL6</li>
			<li>BCOR</li>
			<li>BLM</li>
			<li>BMPR1A</li>
			<li>BRAF</li>
			<li>BRCA1</li>
			<li>BRCA2</li>
			<li>BRD4</li>
			<li>BRIP1</li>
			<li>BTG2</li>
			<li>CASP8</li>
			<li>CBL</li>
			<li>CCND1</li>
			<li>CCND3</li>
			<li>CD274</li>
			<li>CD70</li>
			<li>CDC27</li>
			<li>CDH1</li>
			<li>CDK12</li>
			<li>CDK4</li>
			<li>CDK6</li>
			<li>CDKN1A</li>
			<li>CDKN1B</li>
			<li>CDKN2A</li>
			<li>CDKN2C</li>
			<li>CHD1</li>
			<li>CHD3</li>
			<li>CHD4</li>
			<li>CHEK1</li>
			<li>CHEK2</li>
			<li>CIC</li>
			<li>COL5A1</li>
			<li>CREBBP</li>
			<li>CRLF2</li>
			<li>CSDE1</li>
			<li>CTCF</li>
			<li>CTLA4</li>
			<li>CTNNB1</li>
			<li>CUL3</li>
			<li>CYLD</li>
			<li>CYSLTR2</li>
			<li>DAXX</li>
			<li>DDX3X</li>
			<li>DICER1</li>
			<li>DIS3</li>
			<li>DNAJB1</li>
			<li>DNMT1</li>
			<li>DNMT3B</li>
			<li>EGFR</li>
			<li>EIF1AX</li>
			<li>EIF4A2</li>
			<li>ELF3</li>
			<li>EP300</li>
			<li>EPAS1</li>
			<li>EPCAM</li>
			<li>EPHA3</li>
			<li>EPHA7</li>
			<li>EPHB1</li>
			<li>ERBB2</li>
			<li>ERBB3</li>
			<li>ERBB4</li>
			<li>ERCC2</li>
			<li>ERCC3</li>
			<li>ERCC4</li>
			<li>ERG</li>
			<li>ERRFI1</li>
			<li>ESR1</li>
			<li>ETV6</li>
			<li>EZH2</li>
			<li>FAM58A</li>
			<li>FANCA</li>
			<li>FANCC</li>
			<li>FANCM</li>
			<li>FAT1</li>
			<li>FBXW7</li>
			<li>FGFR1</li>
			<li>FGFR2</li>
			<li>FGFR3</li>
			<li>FGFR4</li>
			<li>FH</li>
			<li>FOXA1</li>
			<li>FOXL2</li>
			<li>FOXO1</li>
			<li>FOXP1</li>
			<li>FUBP1</li>
			<li>GABRA6</li>
			<li>GATA2</li>
			<li>GATA3</li>
			<li>GLI1</li>
			<li>GNA11</li>
			<li>GNA13</li>
			<li>GNAQ</li>
			<li>GNAS</li>
			<li>GPS2</li>
			<li>GREM1</li>
			<li>GSK3B</li>
			<li>GTF2I</li>
			<li>H3F3A</li>
			<li>H3F3C</li>
			<li>HGF</li>
			<li>HIST1H1C</li>
			<li>HIST1H1E</li>
			<li>HIST1H3B</li>
			<li>HIST1H3C</li>
			<li>HIST1H3H</li>
			<li>HLA-A</li>
			<li>HLA-B</li>
			<li>HLA-C</li>
			<li>HNF1A</li>
			<li>HRAS</li>
			<li>IDH1</li>
			<li>IDH2</li>
			<li>IGF1R</li>
			<li>IL6ST</li>
			<li>INPPL1</li>
			<li>IRF2</li>
			<li>IRF4</li>
			<li>IRS2</li>
			<li>JAK1</li>
			<li>JAK2</li>
			<li>JUN</li>
			<li>KDM5C</li>
			<li>KDM6A</li>
			<li>KDR</li>
			<li>KEAP1</li>
			<li>KEL</li>
			<li>KIT</li>
			<li>KLF4</li>
			<li>KMT2A</li>
			<li>KMT2B</li>
			<li>KMT2C</li>
			<li>KMT2D</li>
			<li>KNSTRN</li>
			<li>KRAS</li>
			<li>LATS1</li>
			<li>LATS2</li>
			<li>LYN</li>
			<li>LZTR1</li>
			<li>MAP2K1</li>
			<li>MAP2K2</li>
			<li>MAP2K4</li>
			<li>MAP3K1</li>
			<li>MAP3K13</li>
			<li>MAPK1</li>
			<li>MAX</li>
			<li>MDC1</li>
			<li>MED12</li>
			<li>MEF2B</li>
			<li>MEN1</li>
			<li>MET</li>
			<li>MGA</li>
			<li>MITF</li>
			<li>MLH1</li>
			<li>MLH3</li>
			<li>MRE11</li>
			<li>MSH2</li>
			<li>MSH3</li>
			<li>MSH6</li>
			<li>MST1</li>
			<li>MTOR</li>
			<li>MUTYH</li>
			<li>MYC</li>
			<li>MYCN</li>
			<li>MYOD1</li>
			<li>NBN</li>
			<li>NCOA3</li>
			<li>NCOR1</li>
			<li>NF1</li>
			<li>NF2</li>
			<li>NFE2L2</li>
			<li>NKX3-1</li>
			<li>NOTCH1</li>
			<li>NOTCH2</li>
			<li>NRAS</li>
			<li>NTRK1</li>
			<li>NTRK3</li>
			<li>NUP93</li>
			<li>PAK7</li>
			<li>PALB2</li>
			<li>PARP1</li>
			<li>PAX5</li>
			<li>PBRM1</li>
			<li>PDCD1</li>
			<li>PDGFRA</li>
			<li>PDPK1</li>
			<li>PGR</li>
			<li>PHOX2B</li>
			<li>PIK3CA</li>
			<li>PIK3CB</li>
			<li>PIK3CD</li>
			<li>PIK3CG</li>
			<li>PIK3R1</li>
			<li>PIK3R2</li>
			<li>PIM1</li>
			<li>PLCB4</li>
			<li>PLCG1</li>
			<li>PLK2</li>
			<li>PMS1</li>
			<li>PMS2</li>
			<li>POLD1</li>
			<li>POLE</li>
			<li>POLQ</li>
			<li>POT1</li>
			<li>PPM1D</li>
			<li>PPP2R1A</li>
			<li>PPP4R2</li>
			<li>PPP6C</li>
			<li>PRDM14</li>
			<li>PREX2</li>
			<li>PRKAR1A</li>
			<li>PRKCI</li>
			<li>PRSS1</li>
			<li>PSIP1</li>
			<li>PTCH1</li>
			<li>PTEN</li>
			<li>PTPRD</li>
			<li>PTPRS</li>
			<li>PTPRT</li>
			<li>RAC1</li>
			<li>RAD21</li>
			<li>RAD50</li>
			<li>RAD51</li>
			<li>RAD51B</li>
			<li>RAD51C</li>
			<li>RAD51D</li>
			<li>RAD52</li>
			<li>RAF1</li>
			<li>RASA1</li>
			<li>RB1</li>
			<li>RBM10</li>
			<li>RECQL</li>
			<li>RECQL4</li>
			<li>RET</li>
			<li>RHEB</li>
			<li>RHOA</li>
			<li>RICTOR</li>
			<li>RIT1</li>
			<li>RNF43</li>
			<li>RPL22</li>
			<li>RPL5</li>
			<li>RPS6KA4</li>
			<li>RPTOR</li>
			<li>RRAS2</li>
			<li>RXRA</li>
			<li>SDHA</li>
			<li>SDHAF2</li>
			<li>SDHB</li>
			<li>SDHC</li>
			<li>SDHD</li>
			<li>SESN2</li>
			<li>SETD2</li>
			<li>SMAD2</li>
			<li>SMAD3</li>
			<li>SMAD4</li>
			<li>SMARCA1</li>
			<li>SMARCA4</li>
			<li>SMARCB1</li>
			<li>SMARCD1</li>
			<li>SMC1A</li>
			<li>SMC3</li>
			<li>SMO</li>
			<li>SOCS1</li>
			<li>SOS1</li>
			<li>SOX17</li>
			<li>SOX9</li>
			<li>SPEN</li>
			<li>SPOP</li>
			<li>SPRED1</li>
			<li>SPTA1</li>
			<li>STAG2</li>
			<li>STAT3</li>
			<li>STK11</li>
			<li>STK19</li>
			<li>SUZ12</li>
			<li>TAF1</li>
			<li>TBL1XR1</li>
			<li>TBX3</li>
			<li>TCEB1/ELOC</li>
			<li>TCF12</li>
			<li>TCF3</li>
			<li>TCF7L2</li>
			<li>TET1</li>
			<li>TGFBR1</li>
			<li>TGFBR2</li>
			<li>TLR4</li>
			<li>TNFRSF14</li>
			<li>TP53</li>
			<li>TP63</li>
			<li>TRAF3</li>
			<li>TRAF7</li>
			<li>TSC1</li>
			<li>TSC2</li>
			<li>U2AF1</li>
			<li>USP9X</li>
			<li>VHL</li>
			<li>WT1</li>
			<li>ZFHX3</li>
			<li>ZMYM2</li>
			<li>ZMYM3</li>
		</ul>
	</section>
	<hr class="hz-line">
	<section class="gene-panel">
		<h5 class="font-weight-bold">Intronic and intergenic regions sequenced to increase sensitivity to detect copy-number alterations</h5>
		<ul>
			<li>AKT2</li>
			<li>ALK</li>
			<li>AMER1</li>
			<li>APC</li>
			<li>APOBEC3B</li>
			<li>AR</li>
			<li>AR_ENHANCER</li>
			<li>ARID1A</li>
			<li>ATM</li>
			<li>AXIN1</li>
			<li>BRCA1</li>
			<li>BRCA2</li>
			<li>CCND1</li>
			<li>CCNE1</li>
			<li>CD274</li>
			<li>CDK12</li>
			<li>CDK4</li>
			<li>CDKN2A</li>
			<li>CDKN2B</li>
			<li>CDKN2C</li>
			<li>CHD1</li>
			<li>CHEK2</li>
			<li>CSMD1</li>
			<li>DROSHA</li>
			<li>EGFR</li>
			<li>ERBB2</li>
			<li>FANCA</li>
			<li>FANCC</li>
			<li>FANCD2</li>
			<li>FBXW7</li>
			<li>FGFR1</li>
			<li>FLT3</li>
			<li>FLT4</li>
			<li>FOXA1</li>
			<li>GPC3</li>
			<li>IGF1R</li>
			<li>INTS4</li>
			<li>JUN</li>
			<li>KDM6A</li>
			<li>KIT</li>
			<li>KMT2C</li>
			<li>KRAS</li>
			<li>LRP1B</li>
			<li>MAP2K4</li>
			<li>MCL1</li>
			<li>MDM2</li>
			<li>MDM4</li>
			<li>MEN1</li>
			<li>MET</li>
			<li>MITF</li>
			<li>MLH1</li>
			<li>MSH2</li>
			<li>MYC</li>
			<li>MYCL</li>
			<li>MYCN</li>
			<li>NCOA3</li>
			<li>NF1</li>
			<li>NF2</li>
			<li>NKX2</li>
			<li>NKX3</li>
			<li>NTRK1</li>
			<li>PARK2</li>
			<li>PAX5</li>
			<li>PBRM1</li>
			<li>PDE4D</li>
			<li>PDGFRA</li>
			<li>PIK3CA</li>
			<li>PIK3R1</li>
			<li>PPM1D</li>
			<li>PTEN</li>
			<li>RAF1</li>
			<li>RB1</li>
			<li>SKP2</li>
			<li>SMAD4</li>
			<li>SMARCB1</li>
			<li>SOX2</li>
			<li>STK11</li>
			<li>SUFU</li>
			<li>T_E_DEL</li>
			<li>TERT</li>
			<li>TP53</li>
			<li>TSC1</li>
			<li>TSC2</li>
			<li>VHL</li>
			<li>WT1</li>
			<li>WWOX</li>
			<li>ZBTB16</li>
		</ul>
	</section>
	<hr class="hz-line">
	<section class="gene-panel">
		<h5 class="font-weight-bold">Intronic regions sequenced to detect fusions and structural rearrangements</h5>
		<ul>
			<li>ALK</li>
			<li>AR</li>
			<li>ATM</li>
			<li>BRAF</li>
			<li>BRCA1</li>
			<li>BRCA2</li>
			<li>CD274</li>
			<li>CDK12</li>
			<li>EGFR</li>
			<li>ERG</li>
			<li>ETV6</li>
			<li>EWSR1</li>
			<li>FGFR1</li>
			<li>FGFR2</li>
			<li>FGFR3</li>
			<li>GPR126</li>
			<li>KIT</li>
			<li>NTRK1</li>
			<li>NTRK3</li>
			<li>PDGFRA</li>
			<li>PDGFRB</li>
			<li>PTEN</li>
			<li>RB1</li>
			<li>RET</li>
			<li>ROS1</li>
			<li>TERT</li>
			<li>TMPRSS2</li>
			<li>TP53</li>
		</ul>
	</section>
	<hr class="hz-line">
	<section class="gene-panel">
		<h5 class="font-weight-bold">Pharmacogenomic SNPs analysed</h5>
		<ul>
			<li>COMT</li>
			<li>CYP2D6</li>
			<li>DPYD</li>
			<li>FGFR4</li>
			<li>HOXB13</li>
			<li>HSD3B1</li>
			<li>NQO1</li>
			<li>TPMT</li>
			<li>UGT1A1</li>
		</ul>
	</section>
</body>
</html>
