<!DOCTYPE html>
<html>
<head>
	<meta http-equiv="Cache-control" content="no-cache">
	<meta http-equiv="Expires" content="-1">
	<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
	<meta name="pdfkit-page-size" content="Legal"/>
    <meta name="pdfkit-orientation" content="Landscape"/>
	<link rel="stylesheet" href="/static/css/bootstrap.min.css" >
	<style>
		*{
			font-family: 'Roboto', sans-serif;
			font-size: 14px;
		}

		section {
			padding: 10px 0px 5px 10px;
		}

		table { page-break-after:auto }
		tr    { page-break-inside:avoid; page-break-after:auto }
		td    { page-break-inside:avoid; page-break-after:auto }

		.genomic-alteration {
			width: 100%;
		}
		.genomic-alteration h5, .gene-panel h5 {
			padding-bottom: 10px;
			text-transform: uppercase;
		}

		.genomic-alteration table th {
			/* text-align: center; */
			background-color: #d3d3d3;
			color: #000000;
		}

		.genomic-alteration table td {
			font-weight: 500;
		}

		.genomic-alteration table td.tot-title {
			font-weight: 800;
			text-align: center;
			text-transform: uppercase;
		}

		.gene-panel ul {
			list-style: none;
			display: block;
		}
		
		.gene-panel ul li {
			display: inline-block;
			height: 25px;
			width: 50px;
    		line-height: 30px;
			margin: 5px;
			/* background: #189AB4; */
			/* color: #ffffff; */
			background: #657786;
			color: #ffffff;
			font-size: 10px;
			text-align: center;
			font-weight: 600;
		}

		h5{
			font-weight: bold;
			padding-bottom: 5px;
			color: #B43218;
		}

	</style>
</head>
<body>
	<section>
		<h5 class="font-weight-bold">TECHNICAL VALIDATION</h5>
		<table class="table table-bordered">
			<thead>
				<tr>
					<th>NGS</th>
					<th>ctDNA</th>
					<th>GERMLINE</th>
				</tr>
			</thead>
			<tbody>
				<tr>
					<th>LIBRARY PREP KIT</th>
					<td>Kapa Hyperprep</td>
					<td>Kapa Hyperprep</td>
				</tr>
				<tr>
					<th>INPUT DNA (ng)</th>
					<td>100</td>
					<td>100</td>
				</tr>
				<tr>
					<th>HYBRIDISATION CAPTURE</th>
					<td>ProBio (v1.0; 1.5 Mb)</td>
					<td>ProBio (v1.0; 1.5 Mb)</td>
				</tr>
				<tr>
					<th>TOTAL NUMBER OF READS PAIRS</th>
					<td>34 million</td>
					<td>13 million</td>
				</tr>
				<tr>
					<th>MEAN TARGET COVERAGE</th>
					<td>1461x</td>
					<td>629x</td>
				</tr>
				<tr>
					<th>FRACTION DUPLICATES</th>
					<td>0.32</td>
					<td>0.22</td>
				</tr>
			</tbody>
		</table>
	</section>
	<section class="genomic-alteration">
		<h5 class="font-weight-bold">Categorization of genomic alterations detected by the liquid biopsy assay</h5>
		<table class="table table-bordered table-hover">
			<tr>
				<th colspan="2">MUTATIONS</th>
			</tr>
			<tr>
				<td>ALL CODING EXONS</td>
				<td>51 genes</td>
			</tr>
			<tr>
				<td>MUTATIONAL HOTSPOTS</td>
				<td>27 genes</td>
			</tr>
			<tr>
				<th colspan="2">PHARMACOGENETIC VARIANTS</th>
			</tr>
			<tr>
				<td>SNPs</td>
				<td>4 genes</td>
			</tr>
			<tr>
				<th colspan="2">COPY-NUMBER ALTERATIONS</th>
			</tr>
			<tr>
				<td>TILED SNP FOR GENOME-WIDE CNV</td>
				<td>3128 SNPs</td>
			</tr>
			<tr>
				<td>INCREASED TILING FOR CNV RESOLUTION</td>
				<td>20 genes</td>
			</tr>
			<tr>
				<th colspan="2">STRUCTURAL VARIATION</th>
			</tr>
			<tr>
				<td>GENE FUSION BY INTRONIC SEQUENCING</td>
				<td>3 genes</td>
			</tr>
			<tr>
				<td>GENE BODY SEQUENCING</td>
				<td>8 genes</td>
			</tr>
			<tr>
				<th colspan="2">MICROSATELLITE INSTABILITY & TMB</th>
			</tr>
			<tr>
				<td>MICROSATELLITES</td>
				<td>63</td>
			</tr>
			<tr>
				<td>HYPERMUTATION/TMB ESTIMATION</td>
				<td>YES</td>
			</tr>
			<tr>
				<td>HYPERMUTATION/TMB ESTIMATION</td>
				<td>6</td>
			</tr>
			<tr>
				<th colspan="2">DNA REPAIR DEFICIENCY</th>
			</tr>
			<tr>
				<td>DDR-ASSOCIATED GENES</td>
				<td>16</td>
			</tr>
			<tr>
				<td class="tot-title">Total Size</td>
				<td>1.5 Mb</td>
			</tr>

		</table>
	</section>
	<section class="gene-panel">
		<h5 class="font-weight-bold"> Gene Panel</h5>
		<ul>
			<li>AKT1</li>
			<li>APC</li>
			<li>AR</li>
			<li>ARID1A</li>
			<li>ARID2</li>
			<li>ATM</li>
			<li>ATR</li>
			<li>BARD1</li>
			<li>BRAF</li>
			<li>BRCA1</li>
			<li>BRCA2</li>
			<li>BRIP1</li>
			<li>CCND1</li>
			<li>CDH1</li>
			<li>CDK12</li>
			<li>CDK4</li>
			<li>CDK6</li>
			<li>CDKN1A</li>
			<li>CDKN1B</li>
			<li>CDKN2A</li>
			<li>CDKN2B</li>
			<li>CHD1</li>
			<li>CHEK2</li>
			<li>CTNNB1</li>
			<li>CUL3</li>
			<li>DICER1</li>
			<li>DNMT3A</li>
			<li>ERG</li>
			<li>FANCA</li>
			<li>FOXA1</li>
			<li>FOXO1</li>
			<li>HRAS</li>
			<li>IDH1</li>
			<li>JAK1</li>
			<li>KDM6A</li>
			<li>KEAP1</li>
			<li>KMT2A</li>
			<li>KMT2C</li>
			<li>KMT2D</li>
			<li>KRAS</li>
			<li>MED12</li>
			<li>MET</li>
			<li>MGA</li>
			<li>MLH1</li>
			<li>MLH3</li>
			<li>MRE11A</li>
			<li>MSH2</li>
			<li>MSH3</li>
			<li>MSH6</li>
			<li>MYC</li>
			<li>NBN</li>
			<li>NCOR1</li>
			<li>NKX3-1</li>
			<li>NRAS</li>
			<li>PALB2</li>
			<li>PIK3CA</li>
			<li>PIK3CB</li>
			<li>PIK3CD</li>
			<li>PIK3R1</li>
			<li>PIK3R2</li>
			<li>PMS1</li>
			<li>PMS2</li>
			<li>POLD1</li>
			<li>POLE</li>
			<li>PTEN</li>
			<li>RAD50</li>
			<li>RAD51</li>
			<li>RAD51B</li>
			<li>RAD51C</li>
			<li>RAD51D</li>
			<li>RB1</li>
			<li>RNF43</li>
			<li>SETD2</li>
			<li>SF3B1</li>
			<li>SPEN</li>
			<li>SPOP</li>
			<li>TMPRSS2</li>
			<li>TP53</li>
			<li>U2AF1</li>
			<li>XPO1</li>
			<li>ZBTB16</li>
			<li>ZFHX3</li>
			<li>ZMYM3</li>
		</ul>
	</section>

</body>
</html>
